### PR TITLE
Added substr function

### PR DESCRIPTION
**Pull request by habh-11**

### CODE REVIEW

File: internal/template/template_test.go [hunk 1]
```diff
@@ -298,3 +298,23 @@ func TestMakeSlice(t *testing.T) {
...
+		{"sample string", 0, 40, "sample string"},
+		{"sample string", 0, 5, "sampl"},
+		{"sample string", 5, 4, "e st"},
+		{"sample string", 100, 4, ""},
+		{"sample string", 3, 0, ""},
+	}
+	for _, tc := range tests {
+		got := substr(tc.input, tc.start, tc.length)
+		if diff := cmp.Diff(tc.want, got); diff != "" {
+			t.Errorf("substr result differs (-want +got):\n%v", diff)
```
Comment (yeweidaniel) [inline internal/template/template_test.go:317]: Please give each test case a name or print out the arguments used, so its clear which test failed.

File: internal/template/funcmap.go [hunk 2]
```diff
@@ -156,3 +157,13 @@ func schemaDescription(s string) string {
 
 	return fmt.Sprintf(`"%s"`, s)
 }
+
+func substr(s string, start int, length int) string {
+	if start >= len(s) {
+		return ""
+	}
+	if start+length > len(s) {
```
Comment (yeweidaniel) [inline internal/template/funcmap.go:165]: nit: add space before and after the +

File: internal/template/funcmap.go [hunk 3]
```diff
@@ -156,3 +157,13 @@ func schemaDescription(s string) string {
...
+		length = len(s) - start
+	}
+	return s[start : start+length]
```
Comment (yeweidaniel) [inline internal/template/funcmap.go:168]: nit: add space

File: internal/template/funcmap.go [hunk 4]
```diff
@@ -156,3 +157,13 @@ func schemaDescription(s string) string {
 
 	return fmt.Sprintf(`"%s"`, s)
 }
+
```
Comment (yeweidaniel) [inline internal/template/funcmap.go:160]: please add a comment on what this function does

**Review by yeweidaniel — CHANGES_REQUESTED** [review]

File: internal/template/funcmap.go [hunk 5]
```diff
@@ -156,3 +157,16 @@ func schemaDescription(s string) string {
...
+// substr returns a substring that starts at index 'start'
+// and spans 'length' characters (or until the end of the string,
+// whichever comes first).
+func substr(s string, start int, length int) string {
+	if start >= len(s) {
```
Comment (yeweidaniel) [inline internal/template/funcmap.go:165]: nit: lets return (string, error) with an error message saying what the problem is

**Review by yeweidaniel — APPROVED** [review]

File: internal/template/template_test.go [hunk 6]
```diff
@@ -304,26 +304,28 @@ func TestSubstr(t *testing.T) {
...
 
-		if !tcValid {
-			if err == nil {
-				t.Errorf("Expected error but received nil for input:\"%s\", "+
-					"start:%d, length:%d", tc.input, tc.start, tc.length)
-			} else {
-				continue
-			}
+		if err != nil && !tc.wantErr {
+			t.Errorf("received error for valid input:\"%s\", start:%d, "+
```
Comment (yeweidaniel) [inline internal/template/template_test.go:321]: you can use t.Fatalf instead of Errorf + continue, similar to below

Comment (yeweidaniel) [reply internal/template/template_test.go:321]: also generally its good practice to print out what the err is

Comment (habh-11) [reply internal/template/template_test.go]: Done.

**Review by yeweidaniel — APPROVED** [review]

**Pull request merged**